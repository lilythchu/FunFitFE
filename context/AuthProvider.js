import React, {createContext, useContext, useState} from 'react';

const AuthContext = createContext();

const AuthProvider = ({children}) => {
  const [isLoggedIn, setIsLoggedIn] = useState(false);
  const [profile, setProfile] = useState({});
  const [token, setToken] = useState();

  return (
    <AuthContext.Provider
      value={{
        isLoggedIn,
        setIsLoggedIn,
        profile,
        setProfile,
        token,
        setToken,
      }}>
      {children}
    </AuthContext.Provider>
  );
};

export const useLogin = () => useContext(AuthContext);

export default AuthProvider;
