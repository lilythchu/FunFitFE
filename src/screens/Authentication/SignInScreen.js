import React, {useState} from 'react';
import {
  ActivityIndicator,
  View,
  Text,
  Image,
  ScrollView,
  TouchableOpacity,
  TouchableWithoutFeedback,
} from 'react-native';
import {Icon} from 'react-native-elements';
import Logo from '../../../assets/images/login.png';
import CustomInput from '../../components/CustomInput.js';
import CustomButton from '../../components/CustomButton.js';

import {useNavigation} from '@react-navigation/native';
import {useForm} from 'react-hook-form';
import {useLogin} from '../../../context/AuthProvider';
import {EMAIL_REGEX} from '../../../utils/methods';
import globalStyles from '../../../styles/global';
import client from '../../../api/client';

const SignInScreen = () => {
  const navigation = useNavigation();
  const {
    control,
    handleSubmit,
    formState: {errors},
  } = useForm();
  const {setIsLoggedIn, setProfile, setToken} = useLogin();

  const [loading, setLoading] = useState(false);
  const [message, setMessage] = useState('');
  const [dismiss, setDismiss] = useState(true);
  const [viewPassword, setViewPassword] = useState(false);
  const handleViewPassword = () => {
    setViewPassword(!viewPassword);
  };

  const onSignInPressed = credentials => {
    setLoading(true);
    client
      .post('/user/login', credentials)
      .then(response => {
        const token = response.data.token;
        setToken(token);
        client
          .get('/user/me', {headers: {Authorization: `Bearer ${token}`}})
          .then(res => {
            if (res.data.age === undefined) {
              navigation.navigate('ExtraInfo');
            } else {
              setLoading(false);
              setIsLoggedIn(true);
              setProfile(res.data);
            }
          })
          .catch(error => console.log(error));
      })
      .catch(error => {
        setLoading(false);
        setDismiss(false);
        const mes = error.response.data.message;
        if (mes === 'User does not exist' || mes === 'Incorrect password') {
          setMessage(mes);
        } else {
          setMessage('Oops! Something went wrong, try again');
        }
      });
  };

  const onForgotPasswordPressed = () => {
    navigation.navigate('ForgotPassword');
  };

  const onSignUpPressed = () => {
    navigation.navigate('SignUp');
  };

  return (
    <ScrollView
      showsVerticalScrollIndicator={false}
      style={globalStyles.scrollView}>
      <TouchableWithoutFeedback onPress={() => setDismiss(true)}>
        <View style={globalStyles.root}>
          <Image source={Logo} style={globalStyles.logo} resizeMode="contain" />

          <CustomInput
            name="email"
            icon="envelope"
            label="Email"
            placeholder="Email"
            control={control}
            rules={{
              required: 'Email is required',
              pattern: {value: EMAIL_REGEX, message: 'Email is invalid'},
            }}
          />

          <CustomInput
            label="Password"
            icon="lock"
            name="password"
            rightIcon={
              <TouchableOpacity onPress={handleViewPassword}>
                <Icon
                  type={'font-awesome'}
                  name={viewPassword ? 'eye' : 'eye-slash'}
                  color="#424040"
                />
              </TouchableOpacity>
            }
            placeholder="Password"
            secureTextEntry={!viewPassword}
            control={control}
            rules={{
              required: 'Password is required',
              minLength: {
                value: 3,
                message: 'Password should be minimum 3 characters long',
              },
            }}
          />

          <View style={{flexDirection: 'row', paddingVertical: 10}}>
            <View style={{flex: 1, alignItems: 'flex-end'}}>
              <TouchableOpacity onPress={onForgotPasswordPressed}>
                <Text style={globalStyles.link}>Forgot Password?</Text>
              </TouchableOpacity>
            </View>
          </View>

          <View style={{alignItems: 'center'}}>
            <Text style={{color: 'red', fontSize: 16}}>
              {message && !dismiss ? message : ''}
            </Text>
          </View>

          {loading ? (
            <ActivityIndicator
              size="large"
              style={globalStyles.activityIdicator}
            />
          ) : (
            <CustomButton
              title="Sign in"
              onPress={handleSubmit(onSignInPressed)}
            />
          )}

          <View style={globalStyles.textLinkContainer}>
            <Text>
              Don't have an account?{' '}
              <Text style={globalStyles.link} onPress={onSignUpPressed}>
                Sign up
              </Text>
            </Text>
          </View>

          {/* <SocialSignInButtons /> */}
        </View>
      </TouchableWithoutFeedback>
    </ScrollView>
  );
};

export default SignInScreen;
