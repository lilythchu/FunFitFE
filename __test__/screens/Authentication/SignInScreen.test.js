import React from 'react';
import {fireEvent, render, screen, act, waitFor} from '@testing-library/react-native';
import SignInScreen from '../../../src/screens/Authentication/SignInScreen';
import AuthProvider from '../../../context/AuthProvider';

jest.useFakeTimers();

const mockedNavigate = jest.fn();

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

jest.mock("react-hook-form", () => ({
  ...jest.requireActual("react-hook-form"),
  register: jest.fn(),
  handleSubmit: jest.fn(),
}))

const userObject = {
  email: 'lily@gmail.com',
  password: '123456789',
};

describe('Sign In', () => {
  beforeEach(() => {
    render(
      <AuthProvider>
        <SignInScreen />
      </AuthProvider>
    );
  })

  test('renders Sign in screen correctly', () => {
    expect(screen).toMatchSnapshot();    
  });

  test('should display required error when email and password are invalid', async () => {
    fireEvent.press(screen.getByTestId('signinButton'));
    expect(await screen.findByText('Email is required')).toBeTruthy();
    expect(await screen.findByText('Password is required')).toBeTruthy();
  })

  test('should display matching error when email is invalid', async () => {
    fireEvent.changeText(screen.getByTestId('emailInput'), 'abc@');
    fireEvent.press(screen.getByTestId('signinButton'));
    expect(screen.getByTestId('emailInput').props.value).toEqual('abc@');
    expect(await screen.findByText('Email is invalid')).toBeTruthy();
  });

  test('should display matching error when password is inavlid', async () => {
    fireEvent.changeText(screen.getByTestId('pwdInput'), '12');
    fireEvent.press(screen.getByTestId('signinButton'));
    expect(screen.getByTestId('pwdInput').props.value).toEqual('12');
    expect(await screen.findByText('Password should be minimum 3 characters long')).toBeTruthy();
  });

  test('navigate to forgot password screen', () => {
    fireEvent.press(screen.getByTestId('forgotPwd'));
    expect(mockedNavigate).toBeCalledWith('ForgotPassword');
  })

  test('navigate to sign up screen', () => {
    fireEvent.press(screen.getByTestId('signup'));
    expect(mockedNavigate).toBeCalledWith('SignUp');
  });  
})
