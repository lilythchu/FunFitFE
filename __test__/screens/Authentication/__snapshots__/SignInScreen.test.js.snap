// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign In renders Sign in screen correctly 1`] = `
<RCTScrollView
  showsVerticalScrollIndicator={false}
  style={
    Object {
      "backgroundColor": "white",
      "padding": 20,
    }
  }
>
  <View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={Object {}}
    >
      <Image
        resizeMode="contain"
        source={1}
        style={
          Object {
            "alignSelf": "center",
            "height": 400.2,
            "width": "70%",
          }
        }
      />
      <View>
        <View
          style={
            Object {
              "backgroundColor": "white",
              "paddingHorizontal": 10,
              "width": "100%",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#86939e",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            Email
          </Text>
          <View
            collapsable={false}
            style={
              Object {
                "alignItems": "center",
                "borderBottomWidth": 1,
                "borderColor": "#86939e",
                "flexDirection": "row",
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "height": 24,
                  "justifyContent": "center",
                  "marginRight": 8,
                  "marginVertical": 4,
                  "paddingRight": 4,
                  "width": 24,
                }
              }
            >
              <View
                style={
                  Object {
                    "overflow": "hidden",
                  }
                }
              >
                <View>
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "transparent",
                        "justifyContent": "center",
                      }
                    }
                  >
                    <Text />
                  </View>
                </View>
              </View>
            </View>
            <TextInput
              editable={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              placeholder="Email"
              placeholderTextColor="#86939e"
              style={
                Object {
                  "color": "#242424",
                  "flex": 1,
                  "fontSize": 18,
                  "minHeight": 40,
                }
              }
              testID="emailInput"
              underlineColorAndroid="transparent"
            />
          </View>
          <Text
            style={
              Object {
                "color": "#ff190c",
                "fontSize": 12,
                "margin": 5,
              }
            }
          />
        </View>
      </View>
      <View>
        <View
          style={
            Object {
              "backgroundColor": "white",
              "paddingHorizontal": 10,
              "width": "100%",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#86939e",
                "fontSize": 16,
                "fontWeight": "bold",
              }
            }
          >
            Password
          </Text>
          <View
            collapsable={false}
            style={
              Object {
                "alignItems": "center",
                "borderBottomWidth": 1,
                "borderColor": "#86939e",
                "flexDirection": "row",
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "height": 24,
                  "justifyContent": "center",
                  "marginRight": 8,
                  "marginVertical": 4,
                  "paddingRight": 4,
                  "width": 24,
                }
              }
            >
              <View
                style={
                  Object {
                    "overflow": "hidden",
                  }
                }
              >
                <View>
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "transparent",
                        "justifyContent": "center",
                      }
                    }
                  >
                    <Text />
                  </View>
                </View>
              </View>
            </View>
            <TextInput
              editable={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              placeholder="Password"
              placeholderTextColor="#86939e"
              secureTextEntry={true}
              style={
                Object {
                  "color": "#242424",
                  "flex": 1,
                  "fontSize": 18,
                  "minHeight": 40,
                }
              }
              testID="pwdInput"
              underlineColorAndroid="transparent"
            />
            <View
              style={
                Object {
                  "alignItems": "center",
                  "height": 40,
                  "justifyContent": "center",
                  "marginVertical": 4,
                  "paddingRight": 4,
                }
              }
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "overflow": "hidden",
                    }
                  }
                >
                  <View>
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "backgroundColor": "transparent",
                          "justifyContent": "center",
                        }
                      }
                    >
                      <Text />
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
          <Text
            style={
              Object {
                "color": "#ff190c",
                "fontSize": 12,
                "margin": 5,
              }
            }
          />
        </View>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "paddingVertical": 10,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "flex-end",
              "flex": 1,
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="forgotPwd"
          >
            <Text
              style={
                Object {
                  "color": "#003060",
                }
              }
            >
              Forgot Password?
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
          }
        }
      >
        <Text
          style={
            Object {
              "color": "red",
              "fontSize": 16,
            }
          }
        >
          
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "borderRadius": 3,
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 3,
            },
            Object {
              "marginTop": 10,
            },
            false,
          ]
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="signinButton"
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#68bbe3",
                "borderColor": "transparent",
                "borderRadius": 30,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 8,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "white",
                  "fontSize": 18,
                  "fontWeight": "bold",
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              Sign in
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "marginBottom": 30,
            "padding": 10,
          }
        }
      >
        <Text>
          Don't have an account?
           
          <Text
            onPress={[Function]}
            style={
              Object {
                "color": "#003060",
              }
            }
            testID="signup"
          >
            Sign up
          </Text>
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
